src/SourceMap.js

```diff
@@ -79,12 +79,14 @@ class SourceMap {
   }
 
   addMapping(mapping, lineOffset = 0, columnOffset = 0) {
-    mapping.generated = {
-      line: mapping.generated.line + lineOffset,
-      column: mapping.generated.column + columnOffset
-    };
-
-    this.mappings.push(mapping);
+    this.mappings.push({
+      source: mapping.source,
+      original: mapping.original,
+      generated: {
+        line: mapping.generated.line + lineOffset,
+        column: mapping.generated.column + columnOffset
+      }
+    });
   }
 
   addConsumerMapping(mapping, lineOffset = 0, columnOffset = 0) {
```

Comment:
FYI, the handling below here seems questionable. Mappings with no original location are still important because the terminate the end of a conceptual range over the generated file by saying that a generated location points at nothing. By discarding mappings with no original location, you're essentially force-extending every generated mapping to instead extend to the end of the line/next original-location mapping.

Reply:
Not quite. Sourcemaps are conceptually ranges over the generated code, but the actual mappings passed from the consumer are single points of data, from one generated line/col to an original line/col (optionally).

It's the fact that the original part is option that is critical here, and that is getting lost with the check in

```
     if (
       !mapping.source ||
       !mapping.originalLine ||
       (!mapping.originalColumn && mapping.originalColumn !== 0)
     ) {
       return;
     }
```

Imagine a generated piece of code like this:

    console.log

If I want to map _just_ the `console` to something in the original JS file, setting no original location on the `.log` part, I'd generate two mappings. One would say

```
{
  generated: {
    line: 0,
    column: 0,
  },
  original: {
    line: 0,
    column: 0,
  },
  source: "input.js",
  name: "console",
}
```
to say that the start of the word "console" (line 0, column 0) maps to the original file, also at line 0 column 0.

So far so good, but nothing sets an end position, so that mapping on its own would also apply to the `.log` part.

That's where non-`source` mappings come in. You'd also include a mapping as

```
{
  generated: {
    line: 0,
    column: 7,
  },
  original: null
  source: null,
  name: null,
}
```
to say that, starting at line 0, column 7, the code maps to nothing in the original file.

By discarding that information, you've essentially made the original line 0 column 0 mapping apply to the whole of `console.log`.

Reply:
Ow I never saw that as an issue but I'll definitely look into it.
Now that you've explained it, I can see it could and will cause issues.

Thank you

Reply:
No problem, I'm happy to answer any other sourcemap questions if you have them. Babel has had it's fair share of questionable mapping logic over the ears too.